server/service/src/invoice_line/outbound_shipment_line/insert/mod.rs

```diff
@@ -77,3 +77,257 @@ where
+                &context,
+                &mock_store_a().id,
+                inline_init(|r: &mut InsertOutboundShipmentLine| {
+                    r.id = "new outbound line id".to_string();
+                    r.invoice_id = mock_outbound_shipment_a_invoice_lines()[0]
+                        .invoice_id
+                        .clone();
+                    r.number_of_packs = 40;
+                    r.stock_line_id = mock_stock_line_a().id.clone();
+                    r.item_id = mock_stock_line_a().item_id.clone();
+                }),
+            ),
+            Err(ServiceError::StockLineAlreadyExistsInInvoice(
+                mock_outbound_shipment_a_invoice_lines()[0].id.clone()
+            ))
+        );
+
+        // ReductionBelowZero
+        assert_eq!(
+            service.insert_outbound_shipment_line(
+                &context,
+                &mock_store_a().id,
+                inline_init(|r: &mut InsertOutboundShipmentLine| {
+                    r.id = "new outbound line id".to_string();
+                    r.invoice_id = mock_outbound_shipment_a_invoice_lines()[0]
+                        .invoice_id
+                        .clone();
+                    r.number_of_packs = 8;
+                    r.stock_line_id = mock_stock_line_si_d()[0].id.clone();
+                    r.item_id = mock_stock_line_a().item_id.clone();
+                }),
+            ),
+            Err(ServiceError::ReductionBelowZero {
+                stock_line_id: mock_stock_line_si_d()[0].id.clone(),
+            })
+        );
+
+        //TODO: DatabaseError, NotThisStoreInvoice, NewlyCreatedLineDoesNotExist
+    }
+
+    #[actix_rt::test]
+    async fn insert_outbound_shipment_line_success() {
+        let (_, connection, connection_manager, _) = setup_all(
+            "insert_outbound_shipment_line_success",
+            MockDataInserts::all(),
+        )
+        .await;
+
+        let service_provider = ServiceProvider::new(connection_manager, "app_data");
+        let context = service_provider.context().unwrap();
+        let service = service_provider.invoice_line_service;
+
+        service
+            .insert_outbound_shipment_line(
+                &context,
+                &mock_store_a().id,
+                inline_init(|r: &mut InsertOutboundShipmentLine| {
+                    r.id = "new outbound line id".to_string();
+                    r.invoice_id = mock_outbound_shipment_a_invoice_lines()[0]
+                        .invoice_id
+                        .clone();
+                    r.stock_line_id = mock_stock_line_si_d()[0].id.clone();
+                    r.item_id = mock_stock_line_si_d()[0].item_id.clone();
+                    r.number_of_packs = 1;
+                    r.total_before_tax = 1.0;
+                    r.total_after_tax = 2.0;
+                }),
+            )
+            .unwrap();
+
+        let outbound_line = InvoiceLineRowRepository::new(&connection)
```

Comment:
would also need to check that stock line was updated correctly.

when invoice is new or allocated, available_number_of_packs on stock_line should be reduced by number_of_packs on invoice_line, when invoice is picked, available_number_of_packs and total_number_of_packs should be reduced by number_of_packs on invoice_line